### PR TITLE
Server side validation

### PR DESCRIPTION
**Pull request by michalstruck**
- I opted to remove permissions for client side inserts/updates using the gql api (for the most part) - reason being that otherwise you could just `curl` malicious data anyway,
- forms are submitted using form actions
- because of the above, we're using the service role for inserts/updates. Apart from adding service role permissions for corresponding crud operations using the gql api, I also added checks to prevent crud operations on disallowed rows e.g. user can only modify apps they own - `web/lib/permissions/`

There are still forms that need to be moved to server side validation, opted not do add them now bc the PR grew too big anyway

### CODE REVIEW

File: web/lib/schema/index.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// a place for common schema definitions that should remain uniform across the app
+import * as yup from "yup";
+
+/** use for long form text */
+export const allowedCommonCharactersRegex =
+  /^[\p{Letter}\p{Mark}\s0-9.,!?'"()-;/+]*$/u;
+
+/** use for titles */
+const allowedTitleCharactersRegex = /^[\p{Letter}\p{Mark}\s0-9]+$/u;
```
Comment (michalstruck) [inline web/lib/schema/index.ts:9]: regex defining short form text

File: web/lib/schema/index.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  .string()
+  .url("Invalid URL")
+  .matches(
+    /^https:\/\/(\w+-)*\w+(\.\w+)+([\/\w\-._/?%&#=]*)?$/,
+    "Must be a valid https URL",
+  );
+
+function noLinks(value: string | undefined) {
+  if (!value) return true;
+  const urlRegex = /(https?:\/\/[^\s]+)/g;
```
Comment (michalstruck) [inline web/lib/schema/index.ts:21]: regex defining any url

File: web/lib/schema/index.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
+// a place for common schema definitions that should remain uniform across the app
+import * as yup from "yup";
+
+/** use for long form text */
+export const allowedCommonCharactersRegex =
+  /^[\p{Letter}\p{Mark}\s0-9.,!?'"()-;/+]*$/u;
```
Comment (michalstruck) [inline web/lib/schema/index.ts]: regex defining long form text

**Comment by ik-work-gh** [comment]
@andy-t-wang Can you review?

**Comment by andy-t-wang** [comment]
yes I will look today. Lmk if this seems inline with security recs too

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/SetupForm/graphql/server/update-setup.graphql [hunk 4]
```diff
@@ -5,7 +5,6 @@ mutation UpdateSetup(
   $associated_domains: [String!]
   $contracts: [String!]
   $permit2_tokens: [String!]
-  $can_import_all_contacts: Boolean
```
Comment (andy-t-wang) [inline web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/SetupForm/graphql/server/update-setup.graphql:8]: Why are we getting rid of this?

Comment (andy-t-wang) [reply web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/SetupForm/graphql/server/update-setup.graphql:8]: I guess we might not have the logic right now

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/server/submit.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      contracts,
+      permit2_tokens,
+      whitelisted_addresses,
+    });
+  } catch (error) {
+    if (error instanceof Error && error.message === "Invalid permissions") {
+      throw error;
+    }
+
+    console.log("validateAndUpdateSetupServerSide - error updating setup", {
```
Comment (andy-t-wang) [inline web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/server/submit.ts:68]: mb should be warn?

Comment (andy-t-wang) [reply web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/server/submit.ts:68]: We do want to trigger on call alerts at some point

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/AppTopBar/SubmitAppModal/index.tsx [hunk 6]
```diff
@@ -103,20 +106,14 @@ export const SubmitAppModal = (props: SubmitAppModalProps) => {
         setOpen(false);
       } catch (error: any) {
         console.error("Submit App Modal Failed: ", error);
-        let errorMessage = "Failed to submit app for review";
```
Comment (andy-t-wang) [inline web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/AppTopBar/SubmitAppModal/index.tsx:106]: Why was this removed?

Comment (andy-t-wang) [reply web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/AppTopBar/SubmitAppModal/index.tsx:106]: Knowing they're missing localisation is good to know

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/BasicInformation/index.tsx [hunk 7]
```diff
@@ -158,7 +128,7 @@ export const BasicInformation = (props: {
             placeholder="Enter your App Name"
             maxLength={50}
             addOnRight={
-              <RemainingCharacters text={watch("name")} maxChars={50} />
+              <RemainingCharacters text={watch("name")} maxChars={40} />
```
Comment (andy-t-wang) [inline web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/BasicInformation/index.tsx:131]: why did you change this to 40?

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/BasicInformation/server/submit.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      schema,
+      value: input,
+    });
+
+    if (!isValid || !parsedInput) {
+      throw new Error("Invalid input");
+    }
+
+    const client = await getAPIServiceGraphqlClient();
+    console.log(app_metadata_id, input);
```
Comment (andy-t-wang) [inline web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/BasicInformation/server/submit.ts:26]: why are we logging this?

**Review by andy-t-wang — APPROVED** [review]

**Comment by andy-t-wang** [comment]
This is actually a huge improvement to our overall dev portal structure!

**Comment by andy-t-wang** [comment]
Do not merge till @ottodimi is done with his proxy stuff. We need to heavy QA this in staging

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/SetupForm/graphql/server/update-setup.graphql (hunk 4 above)
Comment (michalstruck) [reply web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/SetupForm/graphql/server/update-setup.graphql:8]: yeah no logic so might as well

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/server/submit.ts (hunk 5 above)
Comment (michalstruck) [reply web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/Advanced/page/server/submit.ts:68]: yeah that was my consideration and actually it's now inconsistent between `submit.ts` files, I'll come up with a logging pattern that makes sense

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/AppTopBar/SubmitAppModal/index.tsx (hunk 6 above)
Comment (michalstruck) [reply web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/AppTopBar/SubmitAppModal/index.tsx:106]: We're letting them know a couple lines up with `Localisation not set for all languages`, and it's not possible to access error messages from form actions, so this would never fire anyway

File: web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/BasicInformation/server/submit.ts (hunk 8 above)
Comment (michalstruck) [reply web/scenes/Portal/Teams/TeamId/Apps/AppId/Configuration/BasicInformation/server/submit.ts:26]: leftover
